server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.concurrent.OpenSearchRejectedExecutionException;
+import org.opensearch.index.shard.IndexShard;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.indices.IndicesService;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Service responsible for applying backpressure for lagging behind replicas when Segment Replication is enabled.
+ *
+ * @opensearch.internal
+ */
+public class SegmentReplicationPressureService {
+
+    private volatile boolean isSegmentReplicationBackpressureEnabled;
+    private volatile int maxCheckpointsBehind;
+    private volatile double maxAllowedStaleReplicas;
+    private volatile TimeValue maxReplicationTime;
+
+    private static final Logger logger = LogManager.getLogger(SegmentReplicationPressureService.class);
+
+    public static final Setting<Boolean> SEGMENT_REPLICATION_INDEXING_PRESSURE_ENABLED = Setting.boolSetting(
+        "index.segrep.pressure.enabled",
+        false,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    public static final Setting<Integer> MAX_INDEXING_CHECKPOINTS = Setting.intSetting(
+        "index.segrep.pressure.checkpoint.limit",
+        4,
+        1,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
```

Comment:
Should this (and other) be index scoped `IndexScope` ?

Reply:
This was my thinking initially, but felt it would get a bit difficult to manage.  I think we can start with node scope and extend to index if the need is there?

Reply:
May be we can rename the setting constants here to reflect here cluster or node scope. 
`index.segrep.pressure.checkpoint.limit` -> `node.segrep....` ?

Reply:
ack good catch - I've just removed the index. prefix.  ex. `segrep.pressure.checkpoint.limit`